### DISCUSSION TITLE
extract `lastModified` property from zip entries

### DISCUSSION THREAD

**Discussion by bwkam** [gildas-lormeau/zip.js#462] [Q&A] [answered]
Hi, I recently discovered this library and it seems pretty good. I was just wondering if I can access the [lastModified](https://developer.mozilla.org/en-US/docs/Web/API/File/lastModified) property from files in the zip because I need that info in my app. 

Thanks.

**Comment by gildas-lormeau** [comment, marked as answer]
The last modification date is accessible via the [`Entry#lastModDate`](https://gildas-lormeau.github.io/zip.js/api/interfaces/Entry.html#lastModDate) property.  See the example below.

```html
<!doctype html>
<html>

<head>
  <title>Display last modification dates of entries in a zip file</title>
  <style>
    table {
      margin-block-start: 8px;
      border-collapse: collapse;
    }
    td {
      border: 1px dotted grey;
      padding: 4px;
    }
  </style>
</head>

<body>
  <form>
    <input type=file id=zipFileInput>
  </form>
  <div id=result>
  </div>
  <script type=module>

import { ZipReader, BlobReader } from "https://unpkg.com/@zip.js/zip.js/index.js";

zipFileInput.onchange = () => display().catch(error => alert(error));

async function display() {
  const zipReader = new ZipReader(new BlobReader(zipFileInput.files[0]));
  const entries = await zipReader.getEntries();
  const tableElement = document.createElement("table");
  entries.forEach(entry => {
    const rowElement = document.createElement("tr");
    const filenameElement = document.createElement("td");
    const lastModDateElement = document.createElement("td");
    filenameElement.textContent = entry.filename;
    lastModDateElement.textContent = entry.lastModDate;
    rowElement.appendChild(filenameElement);
    rowElement.appendChild(lastModDateElement);
    tableElement.appendChild(rowElement);
  });
  result.innerHTML = "";
  result.appendChild(tableElement);
}

  </script>
</body>

</html>
```
Run it on plnkr: https://plnkr.co/edit/X1MwY0hbk2aCxOcM

**Reply by bwkam** [reply]
Thank you so much.

**Reply by gildas-lormeau** [reply]
You're welcome!

**Reply by bwkam** [reply]
Is there a way to change the encoding method used by `Entry.getData()` (should I open a new thread instead?)

**Reply by gildas-lormeau** [reply]
What do you mean by "encoding"? Can you give more details?

**Reply by bwkam** [reply]
Basically, the zip I'm reading contains excel (xlsx) files that I read (with `entry.getData()`). Problem is that when I read the files, they aren't properly encoded like when I read them directly from a file input* (I send the excel's content to an api endpoint therefore the server has a problem parsing them too), I'm not sure why (xlsx isn't utf8 encoded?)

  \*
```js
   fileInput.addEventListener("change", async (e) => {
        const files = mainFileInput.files;
        const file = files[0];
        // send `file` to the server
    });
   ```
   
   ```js
         getZipButton.addEventListener("change", async (e) => {
                const zipReader = new ZipReader(new BlobReader(e.target.files[0]));
                const entries = await zipReader.getEntries();
                entries.forEach(async (entry) => {
                        let id = makeid(10);
                        let writer = new TextWriter();
                        var data = await entry.getData(writer);
                       const blob = new Blob([await entry.getData(writer)], {
                        type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
                       });
                       const file = new File([blob], entry.filename, {
                         type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
                  });
               formData.append(id, file);
          });
   ```
   
   Above is how I read the files from the zip.

**Reply by gildas-lormeau** [reply]
You can pass the encoding of the text as a parameter to the `TextWriter` constructor, see https://gildas-lormeau.github.io/zip.js/api/classes/TextWriter.html#constructor.new_TextWriter. Otherwise, just don't use the `TextWriter` class and use for example the `BlobWriter` class instead. Thus, the extracted file will be handled as a binary file.

**Reply by bwkam** [reply]
That did the trick, thanks again. One last problem with the date, so I get the `rawLastModDate` which is a UNIX timestamp afaik and make that a `Date`, but the date seems wrong.
```js
 let rawDate = await entry.rawLastModDate;
 const date = new Date(rawDate);
 ```
 After getting the year, month..etc, the final date string is `1970-01-18 01:21` which is wrong. 
 Correct is `2023-11-08 16:14`.

**Reply by gildas-lormeau** [reply]
The `rawLastModDate` property returns the date as specified in the ZIP specification, see part 4.4.6 here: https://pkware.cachefly.net/webdocs/APPNOTE/APPNOTE-6.3.10.TXT. The JS code below is the one used to decode it in zip.js. 

https://github.com/gildas-lormeau/zip.js/blob/2a10dda82e7961a7c73e996edf25286a23f1656b/lib/core/zip-reader.js#L682-L689

zip.js can also retrieve this date from the NTFS extra field (see part 4.5.5) or the Unix extra field (see part 4.5.7). See the code below for the implementation in zip.js.

https://github.com/gildas-lormeau/zip.js/blob/2a10dda82e7961a7c73e996edf25286a23f1656b/lib/core/zip-reader.js#L583-L620

https://github.com/gildas-lormeau/zip.js/blob/2a10dda82e7961a7c73e996edf25286a23f1656b/lib/core/zip-reader.js#L691-L693

https://github.com/gildas-lormeau/zip.js/blob/2a10dda82e7961a7c73e996edf25286a23f1656b/lib/core/zip-reader.js#L622-L654

**Reply by bwkam** [reply]
ah I see interesting, I just read the parts in the specification, had no idea date was encoded like this, but the `getDate` function above works like a charm :) Thanks!
